### PR TITLE
refactor router_bgp vrf address-families

### LINKED ISSUE #696: eos_cli_config_gen router bgp data revisit data model
## Issue Type

- [ ] Bug Report
- [x] EOS Configuration rendering error
- [ ] Documentation Issue Report

## Summary

Need to revisit the data model for router bgp vrf definitions in eos_cli_config_gen.

The data model currently provides some abstraction for certain commands and needs to be refactored so that there is no abstraction.

Need to create a dedicated` address-family` section under router bgp vrf definition and support ipv4 and ipv6.
    - `neighbor a.b.c.d activate` should be define under correct address family.
    - `network` should be defined under the address family

The current implementation works, however, it abstracts some of the configurations and doesn't generate matching cli output as per below:

**Current data model:**

```
router_bgp:
  vrfs:
## Tenant_A ##
    TENANT_A_PROJECT01:
      networks:
        10.0.0.0/8:
        100.64.0.0/10:
      neighbors:
        10.2.3.4:
          address_family:
            - ipv4
```

**Desired data model**

```
router_bgp:
  vrfs:
## Tenant_A ##
    TENANT_A_PROJECT01:
    address_families:
      ipv4:   
        neighbors:
          10.2.3.4:
            activate: true
       networks:
          10.0.0.0/8:
          100.64.0.0/10:
```

**Generated:**

```
vrf TENANT_A_PROJECT01
      network 10.0.0.0/8
      network 100.64.0.0/10
      address-family ipv4
         neighbor 10.2.3.4 activate
```

**Desired:**

```
router bgp 65101
   vrf TENANT_A_PROJECT01
      !
      address-family ipv4
         neighbor 10.2.3.4 activate
         network 10.0.0.0/8
         network 100.64.0.0/10
```

### PR DESCRIPTION
**Pull request by carlbuchmann**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #696 

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

```yaml
## new data model ##
router_bgp:
  vrfs:
    < vrf_name_1 >:
      address_families:
        < address_family >:
          neighbors:
            < neighbor_ip_address >:
              activate: < true | false >
#  + define networks per address family
        networks:
          < prefix_address >:
            route_map: < route_map_name >

## old data model ##
router_bgp:
  vrfs:
    < vrf_name_1 >:
      neighbors:
        < neighbor_ip_address >:
          address_family:
            - < address_family >
````

## How to test
see molecule results eos_cli_config_gen

## Checklist:

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [ ] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

### CODE REVIEW

**Comment by carlbuchmann** [comment]
I will update the documentation, once the new data model approved!

**Review by titom73 — COMMENTED** [review]
LGTM!

File: ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2 [hunk 1]
```diff
@@ -46,8 +49,16 @@
 {%                     if tenants[tenant].vrfs[vrf].bgp_peers[bgp_peer].nodes is arista.avd.defined %}
 {%                         if inventory_hostname in tenants[tenant].vrfs[vrf].bgp_peers[bgp_peer].nodes %}
 {%                             set neighbor_params = tenants[tenant].vrfs[vrf].bgp_peers[bgp_peer] %}
-{%                             if neighbor_params.address_family is not arista.avd.defined %}
-{%                                 do neighbor_params.update({'address_family': ['ipv4']}) %}
+{%                             if neighbor_params.address_family is arista.avd.defined %}
```
Comment (carlbuchmann) [inline ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2]: @ClausHolbechArista @titom73 @onurgashi @emilotep - I refactored keeping the current behavior in eos_designs, but I believe we can further abstract this by not requiring the address_family to be defined at all and using ansible.netcommon.ipv4/ipv6 to detect it automatically -> https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters_ipaddr.html

What do you think?

Comment (titom73) [reply ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2]: Looks good to have AF autodetection for v4/v6

Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2]: Yes, looks good to me.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 [hunk 2]
```diff
@@ -506,5 +493,23 @@ router bgp {{ router_bgp.as }}
 {%             endif %}
       {{ aggregate_address_cli }}
 {%         endfor %}
+{%         for  address_family in router_bgp.vrfs[vrf].address_families | arista.avd.natural_sort %}
+      !
+      address-family {{ address_family }}
+{%             for neighbor in router_bgp.vrfs[vrf].address_families[address_family].neighbors | arista.avd.natural_sort %}
+{%                 set neighbor_cli = "neighbor " ~ neighbor %}
+{%                 if router_bgp.vrfs[vrf].address_families[address_family].neighbors[neighbor].activate is arista.avd.defined(true) %}
+{%                     set neighbor_cli = neighbor_cli ~ " activate" %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: `neighbor x.x.x.x` with nothing else after it is not a valid CLI command. So we should check for activate and only print the full cli if it is set.
```
(config-router-bgp-vrf-hat)#neighbor 1.2.3.4 
% Incomplete command
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 [hunk 3]
```diff
@@ -413,13 +413,6 @@ router bgp {{ router_bgp.as }}
 {%         if router_bgp.vrfs[vrf].timers is arista.avd.defined %}
       timers bgp {{ router_bgp.vrfs[vrf].timers }}
 {%         endif %}
-{%         for network in router_bgp.vrfs[vrf].networks | arista.avd.natural_sort %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: why remove this? It is valid CLI in EOS to have networks defined directly under VRF, and being common for all address-families. We could just add the new part and use that for eos_designs.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 (hunk 2 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: Fixed it for you :)

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 (hunk 3 above)
Comment (carlbuchmann) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: you are correct, I will add this back in!

Comment (carlbuchmann) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: re-applied with commit f283455

**Comment by carlbuchmann** [comment]
@ClausHolbechArista @emilotep @onurgashi
I had a review of the `leaf-tenant-router-bgp-evpn-vrfs.j2` template in `eos_designs` with @mpergament and we both believe the auto route-map should be removed:

-> @ line 59
```jinja2
{% do neighbor_params.update({ 'route_map_out':'RM-' ~ vrf ~ '-' ~ bgp_peer ~ '-SET-NEXT-HOP-OUT' }) %}
{%      if neighbor_params.default_originate is arista.avd.defined %}
{%          if neighbor_params.default_originate.route_map is not arista.avd.defined %}
{%              do neighbor_params.default_originate.update({ 'route_map_out':'RM-' ~ vrf ~ '-' ~ bgp_peer ~ '-SET-NEXT-HOP-OUT' }) %}
{%           endif %}
{%       endif %}
```

Was there a specific use case for this? Can this be removed?

**Comment by ClausHolbechArista** [comment]
This was originally planned for servers/firewalls where most important thing would be outbound routemap setting nexthop to varp ip.

**Comment by carlbuchmann** [comment]
> This was originally planned for servers/firewalls where most important thing would be outbound routemap setting nexthop to varp ip.

@ClausHolbechArista 
ok, so then keep it?
Also what is the best way to state the all eos_cli_config_gen commands are supported as entries?

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md [hunk 4]
```diff
@@ -207,8 +210,6 @@ tenants:
               always: < true | false >
             update_source: < interface >
             ebgp_multihop: < 1-255 >
-            address_family:
-              - < ipv4 | ipv6 >
             # Nodes is required to restrict configuration of BGP neighbors to certain nodes in the network.
             nodes: [ < node_1 >, < node_2> ]
             # Next hop settings can be either ipv4 or ipv6 for one neighbor, this will be applied by a uniquely generated route-map per neighbor.
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md:216]: Add route-map-in, route-map-out and password

File: ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2 (hunk 1 above)
Comment (carlbuchmann) [reply ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-router-bgp-evpn-vrfs.j2]: implemented this approach

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md (hunk 4 above)
Comment (carlbuchmann) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md:216]: this has been added to the documentation.

**Review by ClausHolbechArista — APPROVED** [review]
LGTM!

**Pull request merged**